core/modules/replica/Job.h

```diff
@@ -0,0 +1,482 @@
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+#ifndef LSST_QSERV_REPLICA_JOB_H
+#define LSST_QSERV_REPLICA_JOB_H
+
+/// Job.h declares:
+///
+/// class Job
+/// (see individual class documentation for more information)
+
+// System headers
+#include <atomic>
+#include <memory>
+#include <ostream>
+#include <string>
+#include <vector>
+
+// THird party headers
+#include <boost/asio.hpp>
+
+// Qserv headers
+#include "replica/AddReplicaQservMgtRequest.h"
+#include "replica/Controller.h"
+#include "replica/RemoveReplicaQservMgtRequest.h"
+#include "util/Mutex.h"
+
+// Forward declarations
+
+// This header declarations
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+// Forward declarations
+namespace database {
+namespace mysql {
+class Connection;
+}}
+
+/**
+  * Class Job is a base class for a family of replication jobs within
+  * the master server.
+  */
+class Job
+    :   public std::enable_shared_from_this<Job>  {
+
+public:
+
+    /// The pointer type for instances of the class
+    typedef std::shared_ptr<Job> Ptr;
+
+    /// The pointer type for the database connector which provides a database-specific
+    /// SQL generation services.
+    typedef std::shared_ptr<database::mysql::Connection> SqlGeneratorPtr;
+
+    /// Primary public state of the job
+    enum State {
+
+        /// The job has been constructed, and no attempt to execute it has
+        /// been made.
+        CREATED,
+
+        /// The job is in a progress
+        IN_PROGRESS,
+
+        /// The job is finihed. See extended status for more details
+        /// (the completion status, etc.)
+        FINISHED
+    };
+
+    /// Return the string representation of the primary state
+    static std::string state2string(State state) ;
+
+    /// Refined public sub-state of the job once it's FINISHED as per
+    /// the above defined primary state.
+    enum ExtendedState {
+
+        /// No extended state exists at this time
+        NONE,
+
+        /// The job has been fully implemented
+        SUCCESS,
+
+        /// Problems with job configuration found
+        CONFIG_ERROR,
+
+        /// The job has failed
+        FAILED,
+
+        /// Qserv notification failed
+        QSERV_FAILED,
+
+        /// Qserv reported that the source chunk is in use and couldn't be removed
+        QSERV_IN_USE,
+
+        /// Expired due to a timeout (as per the Configuration)
+        EXPIRED,
+
+        /// Explicitly cancelled on the client-side (similar to EXPIRED)
+        CANCELLED
+    };
+
+    /// @return string representation of the extended state
+    static std::string state2string(ExtendedState state) ;
+
+    /// @return string representation of the combined state
+    static std::string state2string(State state, ExtendedState extendedState) {
+        return state2string(state) + "::" +state2string(extendedState);
+    }
```

Comment:
based on my understanding from previous discussions, I think this may not be thread safe. Probably not catastrophically so, but it may be confusing in some cases.